[JIT] add type refinements for isinstance checks

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27007 Remove unnecessary Node* closures in operator registration
* #26788 [jit] clean up exported source format
* #26787 [jit] Make PythonPrint a class
* #26786 [jit] remove dead LEGACY_PythonPrint
* #26767 [jit] Remove attempToRecoverType
* #26572 [jit] Allow 'Any'  to appear as a type argument.
* #26499 [jit] refactor tryMatchSchema
* **#26271 [JIT] add type refinements for isinstance checks**

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

Differential Revision: [D17412856](https://our.internmc.facebook.com/intern/diff/D17412856)

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

suo commented on test/test_jit.py:
```diff
@@ -6771,6 +6771,27 @@ def test(op, args):
+        self.assertEqual(foo2(None, 4), 0)
+        self.assertEqual(foo2(4, None), 0)
+
+
```
nit, extra spaces

eellison reviewed (commented):
it would be good as a follow up to implement a peephole optimization for this op

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head

zdevito committed:
Update on "[JIT] add type refinements for isinstance checks"

[JIT] add type refinements for isinstance checks

This replaces unchecked_unwrap_optional with unchecked_cast. This
enables the generalization of type refinement so that it works for
isinstance checks as well. This also removes unchecked_unwrap_optional from
code we generate, which is good because it is a hard op to serialize well
since it doesn't directly encode the Optional[T] being unwrapped. In contrast,
unchecked_cast always explicitly lists the type.

gh-metadata: pytorch pytorch 26271 gh/zdevito/111/head
